### PR TITLE
feat: graceful shutdown

### PR DESCRIPTION
**Pull request by csviri**

### CODE REVIEW

**Comment by csviri** [comment]
This might deserve also an end to end test, to showcase, how SIGTERM and SIGKILL works in pods. WDYT @metacosm  ?

**Review by metacosm — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/EventProcessor.java [hunk 1]
```diff
@@ -388,6 +388,12 @@ private ReconcilerExecutor(ResourceID resourceID, ExecutionScope<P> executionSco
 
     @Override
     public void run() {
+      if (!running) {
+        // this is needed for the case when controller stopped, but there is a graceful shutdown
+        // timeout. that should finish the currently executing reconciliations but not the ones
+        // which where submitted but not started yet
+        log.debug("Event processor not running skipping resource processing: {}", resourceID);
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/EventProcessor.java:395]: yes, wonder how the test passed, will check

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java [hunk 2]
```diff
@@ -121,6 +121,7 @@ public HasMetadata clone(HasMetadata object) {
    *
    * @return the number of seconds to wait before terminating reconciliation threads
    */
+  @Deprecated(forRemoval = true)
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java:127]: added

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/EventProcessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/EventProcessor.java:395]: fixed the unit test, was not trivial, wonder if we should refactor this in the future.

**Review by metacosm — COMMENTED** [review]
LGTM apart from question regarding deprecation of termination timeout

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java [hunk 3]
```diff
@@ -121,8 +121,12 @@ public HasMetadata clone(HasMetadata object) {
    * Retrieves the number of seconds the SDK waits for reconciliation threads to terminate before
    * shutting down.
    *
+   * @deprecated use {@link io.javaoperatorsdk.operator.Operator#stop(Duration)} instead. Where the
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java:124]: I would prefer to have a single way to configure this, and it is much simpler with the parameters. Keeping this, would be two way to do the same thing.

**Review by metacosm — APPROVED** [review]

**Pull request merged**